content/client-lib-development-guide/features.textile

```diff
@@ -74,21 +74,32 @@ h3(#restclient). RestClient
+* @(RSA14)@ If Token Auth is the default authentication scheme, an exception will be raised if a token is not provided or there is no means to generate a token. For example, if only the option @useTokenAuth@ is specified, and a @key@ is not provided, then the client library is unable to authenticate or issue a token
+* @(RSA15)@ If Token Auth is being used and additionally @clientId@ has been set in the @ClientOptions@ when the library was instanced, the @clientId@ of the token must match the @clientId@ provided in @ClientOptions@. Following an auth request which uses a @TokenDetails@ or @TokenRequest@ object that contains an incompatible @clientId@, the library should change the connection state to @FAILED@ and emit an error
 * @(RSA5)@ TTL for new tokens is specified in milliseconds and defaults to the REST API default (1 hour)
 * @(RSA6)@ The @capability@, if not specified, defaults to allow all operations permitted for the key associated with the token
-* @(RSA7)@ If a @clientId@ is provided in the options, then the string will be encoded in UTF-8 format before it is sent to Ably
+* @(RSA7)@ @clientId@ and authenticated clients:
+** @(RAS7a)@ If a @clientId@ is present in the authentication token, or in the case of a realtime connection, the endpoint has an explicit @clientId@ param configured, then the client is considered to be authenticated i.e. it has an identity that is implicit in all operations. The following applies to authenticated clients:
+*** @(RAS7a1)@ All operations (such as message publishing or presence) will have an implicit @clientId@. The Ably service automatically updates the @clientId@ attribute (when empty) for all @Message@ and @PresenceMessage@ messages received from that authenticated client, and any messages then published from the Ably service, will have the @clientId@ attribute populated. It is therefore expected that Ably client libraries do not explicitly set the @clientId@ field on messages published when @clientId@ is implicit in the connection or authentication scheme
+*** @(RSA7a2)@ The server will reject the publishing of messages with an explicit @clientId@ that does not match the authenticated client's @clientId@. The connection and channel remains available for further operations
+*** @(RSA7a3)@ The client library raises an exception if the @clientId@ attribute of any @Message@ or @PresenceMessage@ published is not @null@ and does not match the non-@null@ @Auth#clientId@. The connection and channel remains available for further operations
+*** @(RSA7a4)@ @Auth#clientId@ attribute returns a string value for the authenticated client's @clientId@. If the client is not authenticated and a @clientId@ value was not provided when instanced with @ClientOptions@, then @Auth#clientId@ is @null@
+** @(RSA7e)@ @Auth#clientId@ is not @null@ when:
+*** @(RSA7e1)@ A @clientId@ is provided in the @ClientOptions@. @clientId@ should be a UTF-8 string
```

Comment:
> being explicit about Strings being an array of Unicode characters is important

Absolutely, but really this goes without saying. For absolute clarity we should state that clientIds, channel names, event names are all Unicode strings and Unicode support is therefore required for those entities in all libraries. That's not in dispute.

> So if the language forces an encoding, we expect client libraries to use UTF-8 encoding.

In a language like Ruby in which string values are internally tagged with an encoding, then I think it's fine for the API to prefer (or insist), for efficiency reasons, that the caller supplies values tagged (ie encoded) with UTF-8. That is really an issue I think for the Ruby client library documentation.

Any sensible language hides from the developer how strings are internally encoded, except sometimes for artifacts of the encoding visible from the handling of surrogate pairs. It is not correct that we ask for a specific default if the language does expose this; for example Java and Javascript will internally use UCS-2 or UTF-16. If the language does offer the programmer a choice of internal representation for strings, then the library IMO should use the language default. Then it is up to the msgpack or other libraries to ensure compatibility on the wire.